managed private networks: update the create and list commands

mcorbin opened the pull request:
#319 

- Add the `startip`, `endip`, `netmask` parameters to the create
  network command
- Show the `startip`, `endip`, `netmask` fields in the list networks command

greut commented on networks.go:
```diff
@@ -114,18 +116,18 @@ type CreateNetwork struct {
 	IsolatedPVlan     string `json:"isolatedpvlan,omitempty" doc:"the isolated private vlan for this network"`
 	Name              string `json:"name,omitempty" doc:"the name of the network"` // This field is required but might be empty
-	Netmask           net.IP `json:"netmask,omitempty" doc:"the netmask of the network. Required for Shared networks and Isolated networks when it belongs to VPC"`
+	Netmask           net.IP `json:"netmask,omitempty" doc:"the netmask of the network.  Required for managed networks"`
```
```console
$ go run generate/main.go -apis listApis.json -cmd CreateNetwork

netmask: wrong doc want "the netmask of the network. Required for managed networks." got "the netmask of the network.  Required for managed networks"
```

greut commented on networks.go:
```diff
@@ -30,7 +31,7 @@ type Network struct {
 	IsSystem                    bool          `json:"issystem,omitempty" doc:"true if network is system, false otherwise"`
 	Name                        string        `json:"name,omitempty" doc:"the name of the network"`
-	Netmask                     net.IP        `json:"netmask,omitempty" doc:"the network's netmask"`
+	Netmask                     net.IP        `json:"netmask,omitempty" doc:"the network's netmask. Required for managed networks"`
```
this doesn't match the content of the `listApis`

greut commented on networks.go:
```diff
@@ -155,10 +157,13 @@ type UpdateNetwork struct {
+	EndIP             net.IP `json:"endip,omitempty" doc:"the ending IP address in the network IP range. Required for managed networks."`
 	GuestVMCIDR       *CIDR  `json:"guestvmcidr,omitempty" doc:"CIDR for Guest VMs,Cloudstack allocates IPs to Guest VMs only from this CIDR"`
 	Name              string `json:"name,omitempty" doc:"the new name for the network"`
+	Netmask           net.IP `json:"netmask,omitempty" doc:"the netmask of the network.  Required for managed networks"`
```
```console
$ go run generate/main.go -apis listApis.json -cmd UpdateNetwork

netmask: wrong doc want "the netmask of the network. Required for managed networks." got "the netmask of the network.  Required for managed networks"
```

greut commented on cmd/exo/cmd/privnet_create.go:
```diff
@@ -83,12 +96,25 @@ func privnetCreate(name, desc, zoneName string) error {
+	}
+	if netmaskAddr != "" {
+		netmask = net.ParseIP(netmaskAddr)
+	}
```
the `if` checks are not required.

greut commented on cmd/exo/cmd/privnet_create.go:
```diff
@@ -102,15 +128,18 @@ func privnetCreate(name, desc, zoneName string) error {
-	table.SetHeader([]string{"Name", "Description", "ID"})
-	table.Append([]string{newNet.Name, newNet.DisplayText, newNet.ID.String()})
+	table.SetHeader([]string{"Name", "Description", "ID", "StartIP", "EndIP", "Netmask"})
+	table.Append([]string{newNet.Name, newNet.DisplayText, newNet.ID.String(), newNet.StartIP.String(), newNet.EndIP.String(), newNet.Netmask.String()})
```
`<nil>` is an implementation details that shouldn't be shown.

| NAME | DESCRIPTION |                  ID                  | STARTIP | ENDIP | NETMASK |
|------|-------------|--------------------------------------|---------|-------|---------|
| bar  |             | d8283a3a-a9ff-4f5d-b385-f7ad16d4c311 | \<nil>   | \<nil> | \<nil>   |

greut commented on cmd/exo/cmd/privnet_create.go:
```diff
@@ -83,12 +96,19 @@ func privnetCreate(name, desc, zoneName string) error {
 
+	startip := net.ParseIP(startIPAddr)
+	endip := net.ParseIP(endIPAddr)
+	netmask := net.ParseIP(netmaskAddr)
```
:+1:

mcorbin commented:
I rebased by branch.
I removed everything related to updateNetworks (and will move it in my other PR).
`go run generate/main.go -apis listApis.json -cmd CreateNetwork` is now OK.

The tables for createNetwork/listNetworks show empty values instead of <nil>.

mcorbin commented on cmd/exo/cmd/privnet_list.go:
```diff
@@ -59,7 +61,14 @@ func listPrivnets(zone string, table *table.Table) error {
+				vmNum,
+				utils.EmptyStringIfNilIP(pn.StartIP),
+				utils.EmptyStringIfNilIP(pn.EndIP),
+				utils.EmptyStringIfNilIP(pn.Netmask)})
```
Done ;)

greut commented on cmd/exo/cmd/privnet.go:
```diff
@@ -34,3 +37,13 @@ func getNetworkByName(name string) (*egoscale.Network, error) {
 }
+
+// getDhcpString returns the string representation for a DHCP
+func getDhcpString(startIP, endIP, netmask net.IP) string {
```
`s/getDhcpString/dhcpRange/`

https://golang.org/doc/effective_go.html#Getters

mcorbin replied:
done

mcorbin commented:
I updated the privnet create command:

```
exo privnet create foobar --startip 10.0.0.20 --endip 10.0.0.30 --mask 24
┼────────┼─────────────┼──────────────────────────────────────┼─────────────────────────┼
│  NAME  │ DESCRIPTION │                  ID                  │          DHCP           │
┼────────┼─────────────┼──────────────────────────────────────┼─────────────────────────┼
│ foobar │             │ 3180f776-dcb6-42b7-9074-db9f7668a6ee │ 10.0.0.20-10.0.0.30 /24 │
┼────────┼─────────────┼──────────────────────────────────────┼─────────────────────────┼
```

greut commented:
@mcorbin then it matches the doc, thanks.

```go
   Use: "create <name>",
```

vincentbernat commented on cmd/exo/cmd/privnet_create.go:
```diff
@@ -102,15 +116,23 @@ func privnetCreate(name, desc, zoneName string) error {
 	privnetCreateCmd.Flags().StringP("description", "d", "", "Private network description")
+	privnetCreateCmd.Flags().StringP("startip", "s", "", "the beginning IP address in the network IP range. Required for managed networks.")
+	privnetCreateCmd.Flags().StringP("endip", "e", "", "the ending IP address in the network IP range. Required for managed networks.")
+	privnetCreateCmd.Flags().StringP("mask", "m", "", "the netmask of the network.  Required for managed networks")
```
"mask", could be `UintP`?

mcorbin replied:
yes, updated ;) thanks

greut replied:
I'd still accept both though.

mcorbin replied:
You want `mask` and `netmask` ?

mcorbin replied:
Now, the command supports both.

greut replied:
@mcorbin the minimum is to be as good as the API, then we can do more but I wouldn't do things differently.

greut commented on cmd/exo/cmd/privnet.go:
```diff
@@ -34,3 +37,23 @@ func getNetworkByName(name string) (*egoscale.Network, error) {
+	return "n/a"
+}
+
+func netmaskString(ip net.IP, mask uint) net.IP {
```
this is confusing: `func netmaskString -> net.IP`, return a string.

mcorbin replied:
Updated

greut commented on cmd/exo/cmd/privnet_create.go:
```diff
@@ -16,11 +17,27 @@ var privnetCreateCmd = &cobra.Command{
+		if err != nil {
+			return err
+		}
+		netmask, err := cmd.Flags().GetString("netmask")
```
I'd keep only one of them, it's easier to support `255.255.255.0`, `/24` or `24` syntaxes.

mcorbin replied:
Ok, now the command supports only netmask (255.255.255.0)

greut commented on networks.go:
```diff
@@ -114,18 +116,18 @@ type CreateNetwork struct {
 	IsolatedPVlan     string `json:"isolatedpvlan,omitempty" doc:"the isolated private vlan for this network"`
 	Name              string `json:"name,omitempty" doc:"the name of the network"` // This field is required but might be empty
-	Netmask           net.IP `json:"netmask,omitempty" doc:"the netmask of the network. Required for Shared networks and Isolated networks when it belongs to VPC"`
+	Netmask           net.IP `json:"netmask,omitempty" doc:"the netmask of the network. Required for managed networks."`
```
could it be `net.IPMask`?

mcorbin replied:
Nope, the API needs an IP.

greut replied:
we would have to create how own type, right. Not worth it.

greut commented on cmd/exo/cmd/privnet_create.go:
```diff
@@ -35,12 +48,6 @@ var privnetCreateCmd = &cobra.Command{
-				if err != nil {
-					return err
-				}
-			}
```
`if zone == "" {` should be enough

mcorbin replied:
Updated

greut commented on cmd/exo/cmd/privnet_create.go:
```diff
@@ -16,11 +17,23 @@ var privnetCreateCmd = &cobra.Command{
 	Aliases: gCreateAlias,
 	RunE: func(cmd *cobra.Command, args []string) error {
-		name, err := cmd.Flags().GetString("name")
+		if len(args) < 1 {
```
`if len(args) != 1` seems more accurate.

mcorbin replied:
Updated

greut commented on cmd/exo/cmd/privnet_create.go:
```diff
@@ -83,12 +90,19 @@ func privnetCreate(name, desc, zoneName string) error {
 		ZoneID:      zone,
+		StartIP:     startIP,
+		EndIP:       endIP,
+		Netmask:     netmaskIP,
```
You're relying on the API for doing the validation here, which might be okay.
